File: Linked-Lists/circular_linked_list.py
class Node:
    def __init__(self, data: int):
        self.data = data
        self.next = self

class CircularLinkedList:
    def __init__(self):
        self.head = None

    def insert_first(self, data: int) -> None:
        new_node = Node(data)
        if not self.head:
            # if list is empty, new node becomes head
            self.head = new_node
        else:
            # point new_node to last
            new_node.next = self.head
            current_node = self.head
            # traverse list till last node
            while current_node.next != self.head:
                current_node = current_node.next
            # point last node to new node
            current_node.next = new_node
            # make new node current head
            self.head = new_node

    def insert_last(self, data: int) -> None:
        new_node = Node(data)
        current_node = self.head

        # traverse list till the last node
        while current_node.next != self.head:
            current_node = current_node.next
        # point last node to newly created node
        current_node.next = new_node
        # point new node to head thus making it the last node
        new_node.next = self.head

    def insert_between_nodes(self, data: int, prev_node_data: int) -> None:
        new_node = Node(data)
        ref_node = self.search(prev_node_data)
        new_node.next = ref_node.next
        ref_node.next = new_node

    def search(self, data: int) -> Node:
        current_node = self.head
        while current_node.data != data:
            current_node = current_node.next
        return current_node

    def traverse(self) -> None:
        current_node = self.head
        while current_node.next != self.head:
            print(current_node.data)
            current_node = current_node.next
        print(current_node.data)

    def delete_node(self, prev_node_data: int, data: int) -> None:
        ref_node = self.search(prev_node_data)
        deleted_node = self.search(data)
        ref_node.next = deleted_node.next
        del(deleted_node)

if __name__ == '__main__':
    circular_linked_list = CircularLinkedList()
    circular_linked_list.insert_first(5)
    circular_linked_list.insert_first(2)
    circular_linked_list.insert_first(7)
    circular_linked_list.insert_last(6)
    circular_linked_list.insert_last(9)
    circular_linked_list.insert_between_nodes(4, 7)
    circular_linked_list.insert_between_nodes(8, 6)
    # circular_linked_list.delete_node(6, 8)
    circular_linked_list.traverse()

File: Linked-Lists/Circular Linked List/src/CircularLinkedList.java
package src;

public class CircularLinkedList {
    private Node head;
    private Node tail;
    private int size;

    public CircularLinkedList() {
        this.head = null;
        this.tail = null;
        this.size = 0;
    }

    private class Node {
        private int data;
        private Node next;

        private Node(int data) {
            this.data = data;
            this.next = null;
        }
    }

    public boolean isEmpty() {
        return this.size == 0;
    }

    public int getSize() {
        return this.size;
    }

    public void insertFirst(int data) {
        Node newNode = new Node(data);

        if (this.head == null) {
            this.head = newNode;
            this.tail = newNode;
            this.head.next = this.head;
        } else {
            newNode.next = this.head;
            this.head.next = newNode;
            this.head = newNode;
        }

        size++;
    }

    public void insertLast(int data) {
        Node newNode = new Node(data);

        if (this.tail == null) {
            this.head = newNode;
            this.tail = newNode;
            this.head.next = this.head;
        } else {
            this.tail.next = newNode;
            newNode.next = this.head;
            this.tail = newNode;
        }

        size++;
    }

    public void insertItem(int dataInWantedNodeLocation, int newData) {
        /*
        Since I have no indices, dataInWantedNodeLocation helps us
        find the location where we want to insert data to. 
        */
        Node newNode = new Node(newData);
        Node currentNode = this.head;

        while (currentNode.data != dataInWantedNodeLocation) {
            currentNode = currentNode.next;
        }

        newNode.next = currentNode.next;
        currentNode.next = newNode;

    }

    public void traverse() {
        Node currentNode = this.head;
        int nodeNumber = 0;

        while (currentNode.next != this.head) {
            System.out.println(String.format("Node number: %s, data: %s", nodeNumber, currentNode.data));
            currentNode = currentNode.next;
            nodeNumber++;
        }

        //Prints out data in last node to prevent circular motion
        System.out.println(String.format("Node number: %s, data: %s", nodeNumber, currentNode.data));
    }

    public Node deleteHead() {
        Node temp = this.head;

        this.head = this.head.next;
        this.tail.next = this.head; // Make tail next pointer point to new head

        return temp;
    }

    public Node deleteTail() {
        Node currentNode = this.head;
        Node temp = null;

        while (currentNode.next != this.tail) {
            currentNode = currentNode.next;
        }

        temp = currentNode.next;
        currentNode.next = this.head;
        this.tail = currentNode;

        return temp;
    }

    public Node deleteNode(int data) {
        Node currentNode = this.head;

        while (currentNode.next.data != data) {
            currentNode = currentNode.next;
        }

        Node temp = currentNode.next;

        currentNode.next = currentNode.next.next;

        return temp;
    }
}
